### PR TITLE
feat: Implement ingress and egress with prefix lists

### PR DESCRIPTION
**Pull request by andyshinn**
## Description

This is a reopening of pull request #226 which didn't get any movement. I am opening after forking and rebasing to see if we can get some movement on it as I would like the feature.

This feature aims at allowing the module to provision SG Rules by specifying explicitly the ingress or ingress rules with prefix lists. It also allows to unset a new variable enable_prefix_lists_cross_over which drives whether there should be cross over flow openings between the inputs (self, cidr blocks, or security groups) and the prefix lists.

## Motivation and Context

I am creating security groups with prefix lists as ingress and egress targets.

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [ ] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [ ] I have executed `pre-commit run -a` on my pull request

I am still going through the testing parts (trouble getting pre-commit to run). But I am applying this to projects using my fork.

### CODE REVIEW

**Comment by andyshinn** [comment]
Thanks, bot. It is still a desired PR.

**Comment by andyshinn** [comment]
Thanks, bot. It is still a desired PR.

**Comment by andyshinn** [comment]
Thanks, bot. It is still a desired PR.

**Comment by andyshinn** [comment]
Thanks, bot. It is still a desired PR.

**Review by loicvolle — APPROVED** [review]
LGTM

**Comment by keenobi** [comment]
Hello ! 
Thanks for the PR, this is exactly what we are looking for ! :+1: 

@antonbabenko is it something that may block the merge ? 

Best regards

**Comment by andyshinn** [comment]
Thanks, bot. It is still a desired PR.

**Comment by andyshinn** [comment]
Rebased to fix conflicts.

File: examples/complete/main.tf [hunk 1]
```diff
@@ -437,3 +437,30 @@ module "prefix_list" {
...
+module "prefix_list_sg" {
+  source = "../../"
+
+  name        = "prefix-list-sg"
+  description = "Security group using prefix list"
+  vpc_id      = data.aws_vpc.default.id
+
+  use_name_prefix = false
+
+  ingress_prefix_list_ids = [aws_ec2_managed_prefix_list.prefix_list_sg_example.id]
```
Comment (antonbabenko) [inline examples/complete/main.tf:464]: Please use `ingress_with_prefix_list_ids` instead in the example to show the proposed feature for real.

File: main.tf [hunk 2]
```diff
@@ -430,6 +430,67 @@ resource "aws_security_group_rule" "computed_ingress_with_self" {
...
   )
 }
+# Security group rules with "prefix_list_ids", but without "cidr_blocks", "self" or "source_security_group_id"
+resource "aws_security_group_rule" "ingress_with_prefix_list_ids" {
+  count = var.create ? length(var.ingress_with_prefix_list_ids) : 0
+
+  security_group_id = local.this_sg_id
+  type              = "ingress"
+
+  prefix_list_ids = var.ingress_prefix_list_ids
```
Comment (antonbabenko) [inline main.tf:440]:
```suggestion
  prefix_list_ids = var.ingress_with_prefix_list_ids
```
Or something along those lines, right?

Please verify in other resources, too.

File: variables.tf [hunk 3]
```diff
@@ -118,6 +118,12 @@ variable "ingress_prefix_list_ids" {
   default     = []
 }
 
+variable "ingress_with_prefix_list_ids" {
```
Comment (antonbabenko) [inline variables.tf:121]: There should be similar addition to `modules/_templates` to add support to named submodules.

Also, please update an example in `examples/http` to show the usage of this feature.

**Review by antonbabenko — CHANGES_REQUESTED** [review]
Looks rather good, but there are still a few things to do.

PS: Apologies that it took ages to review this PR.

**Comment by w0rldart** [comment]
What is missing for this PR to be merged?

**Comment by andyshinn** [comment]
I will fix the feedback items soon. I am busy last few weeks.

File: variables.tf (hunk 3 above)
Comment (andyshinn) [reply variables.tf:121]: Updated!

File: main.tf (hunk 2 above)
Comment (andyshinn) [reply main.tf:440]: I'm a little confused here. The `ingress_with_prefix_list_ids` is the list of custom rules to apply, not the list of prefix IDs. I was just following the template of the other ones which create a resource per rule (`ingress_with_prefix_list_ids` and then set the argument to the resource ID to apply to (`ingress_prefix_list_ids`). Let me know if I have confused something and if you could clarify.

Comment (antonbabenko) [reply main.tf:440]: You're right. I misunderstood the code.

**Review by antonbabenko — APPROVED** [review]
LGTM

**Comment by antonbabenko** [comment]
Thank you @andyshinn !

**Comment by antonbabenko** [comment]
This PR is included in version 5.1.0 :tada:

**Pull request merged**